### PR TITLE
Fix #246 Switch to syncstorage-rs, and document the additional depend…

### LINKED ISSUE #246: Update syncserver package to use the new syncstorage-rs module.
As Python 2.7 is now end of life, Mozilla has been working toward reimplementing the syncstorage backend in rust. This is now complete. https://github.com/mozilla-services/syncstorage-rs

Now that syncstorage-rs is ready for use, we should update this repo to use it.

### PR DESCRIPTION
**Pull request by fzzzy**
…encies required.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,14 +1,16 @@
 Run-Your-Own Firefox Sync Server
 ================================
 
+Note: For the duration of Q4 2020 this documentation will be revised. This situation will only be temporary until tokenserver is ported to rust and this documentation is rewritten to leave out the old dependencies.
```
Comment (jrconlin) [inline README.rst:4]: May want to emphasize this a bit. maybe 

`**Note:**`
or offset things in a block quote just to make sure folks see that.

File: README.rst [hunk 2]
```diff
@@ -35,6 +37,18 @@ following packages (or similar, depending on your operating system) installed:
...
 - openssl-dev
 
+syncstorage-rs uses the following additional dependencies:
+
+- cmake
+- golang
+- libcurl4-openssl-dev
+- libssl-dev
+- pkg-config
+- Rust stable
```
Comment (jrconlin) [inline README.rst:47]: I wonder if we need to add a link to `rustup.rs` here? Not sure if `rustlang` is available on all the platforms yet. Did a quick search on [Debian buster](https://packages.debian.org/buster/devel/) and could only find the source.

File: syncserver/__init__.py [hunk 3]
```diff
@@ -155,7 +159,9 @@ def includeme(config):
 
     # Include the relevant sub-packages.
     config.scan("syncserver", ignore=["syncserver.wsgi_app"])
-    config.include("syncstorage", route_prefix="/storage")
+
+    subprocess.call("%sSYNC_MASTER_SECRET=%s /usr/bin/env sh run-syncstorage-rs.sh &" % (sync_database_url, secret, ), shell=True)
+
```
Comment (jrconlin) [inline syncserver/__init__.py:164]: Really wish shells were more consistent about how they recognize ENV vars in CLI, but that's a personal thing.

Comment (fzzzy) [reply syncserver/__init__.py:164]: Please, if you have a better way to do this, let's do it. I suppose I could set the environment in the parent, then call the child, instead?

Comment (jrconlin) [reply syncserver/__init__.py:164]: IIRC you can pass `env={"SYNC_MASTER_SECRET": secret, "SYNC_DATABASE_URL": sync_database_url}` as an argument to `subprocess.call()`. I'd have to experiment a bit to verify (I think there are problems on Windows, but I'm not 100% sure we care).

Comment (fzzzy) [reply syncserver/__init__.py:164]: Setting os.environ in the parent before subprocess.call forked the child worked.

**Review by jrconlin — APPROVED** [review]

**Comment by rfk** [comment]
Sorry I didn't get a chance to check this out sooner.

What will the update experience be like here for existing users, particularly those following `master` (like https://github.com/mozilla-services/syncserver/issues/252)? IIUC this will require them to set up a fresh db using MySQL. It's also not clear to me what will happen if their existing config says to use a different db such as sqlite.

**Comment by fzzzy** [comment]
We're intending for old mysql databases to be compatible. If that's not the case we'll consider it a bug in syncstorage-rs.

Yes, you are right that old config options will not be supported. I have fleshed out plans for supporting old config options for the tokenserver in https://docs.google.com/document/d/10pwmjY1zMBzIdNpL3TV2FZXQLi2ste0Y2FQdcDyT4Ug/edit?ts=5f74fa2c#heading=h.lwwz96fhkfsv

If you wish, I can try my best to support old config options for the storage server. Let's file bugs for whatever we think needs supporting. Sqlite support isn't likely any time soon; could we just print an error message in those cases? However if someone does want to make a Sqlite backend, syncstorage-rs uses diesel for the mysql implementation which theoretically also supports sqlite, although I think there is a fair amount of plumbing.

**Comment by rfk** [comment]
Something else for us to consider here is users who are currently using a non-MySQL database for self-hosting, e.g. I believe the default setup is to use a sqlite db.

**Comment by fzzzy** [comment]
I think we'd like to retire the python code before we'll have a rust-based sqlite version. Would printing a deprecation message with a link to inquire if someone encounters that situation be enough? Then we could have whoever inquires because they actually do use a sqlite database test the (to-be-written) rust code that can transparently keep using that same sqlite db.

**Comment by fzzzy** [comment]
We could print a warning and fall back to the old python 2.7 code if someone is configured to use a sqlite database until the rust code is ready, actually. I think we should do that, as long as we anticipate the rust sqlite version being ready within the next 6 months, which I think is (? engineer estimate alert) reasonable.

**Comment by fzzzy** [comment]
I have investigated this more, and it makes sense and is easy to fall back to the python 2.7 implementation if we encounter an option that rust doesn't support. The responsibility is on the person using the configuration option that causes a python 2.7 fallback to make sure they have a working python 2.7 environment with all the right dependencies.

**Pull request merged**